### PR TITLE
Many new features and tweaks

### PR DESCRIPTION
**Pull request by torcado194**
Hi there!

I want to start this off by saying I love this plugin, I think it's entirely more intuitive than the built-in tilemap system. I started off learning godot with version 4.0 almost solely because of the purportedly improved tile system, and I still haven't gotten used to it. After using this plugin for only a day, I've completely grasped how to use it effectively for my needs.

I began these modifications as separate git branches, but it became too much of a hassle to manage and I ended up writing everything at once. If you would like me to split this into different PRs I can try to do so.

# Changes

### Improved painting
Painting now draws unbroken lines when moving the cursor quickly. Here are some before and after examples:

Before | After
:-------:|:-------:
sx-858 | sx-865
sx-874 |sx-872

This also works when painting tile peering sides.
> Note: This doesn't create perfect unbroken lines with tile shapes other than square. There's an alternative algorithm that was written for the new tileset system internally that's used for other tile shapes, I'll look into implementing this later.

### Line paint tool
Selecting the Line tool, or holding Shift with the normal Paint tool active, will draw lines. This mirrors the built-in editor behavior.

sx-878

> Note: like with the improved painting, lines are slightly wrong with non-square tile shapes

### Overwrite painting mode
A new button to the right of the paint tools toggles "overwrite" mode. While this mode is active, drawing will only change tiles on the tilemap if there is a tile with a matching set of peering sides in the current terrain.  
This allows you to easily manage alternative tilesets, such as a tileset with slopes instead of square corners, or a tileset of alt decoration tiles.

sx-880

### Select and Copy/Paste/Delete terrain peering
A new button on the terrain side allows you to select terrain tiles. Pressing Delete will delete the tiles' peering sides. Pressing Ctrl+C or Ctrl+X will copy or cut the tiles, and Ctrl+V will put the tiles in a staging area on your cursor. Clicking will then overwrite any peering data with that of the staged tiles that were copied.  
This allows you to easily duplicate a tileset's peering sides, for instance if you have many tilesets with the same layout, without requiring you to draw it all again.

sx-896

### Ctrl-scroll zoom keybind
You can now hold ctrl while scrolling to zoom the tileset canvas.

### More clear terrain tiles
This change puts an outline around the tiles selected for the current terrain, just to make it more clear what the current terrain tiles are without needing to look for the matching color between the tile centers and the terrain icon. You can see this change in some of the gifs above.

------

These are mostly just changes I wanted for my own sake, but I figured I should make a PR for them :)
Feel free to suggest any changes or additions! There are likely places in the code that could be improved, I tried my best to mimic the existing style and understand the internal systems.

Thanks! ❤️

### CODE REVIEW

**Comment by Portponky** [comment]
Wow, these are some huge QOL improvements. Thanks so much for taking the time to do this. I'll check through the code tonight and give any feedback.

**Comment by Portponky** [comment]
Hi, I've had a look over the code. Great work. There are a few changes I could suggest, or if you would prefer, I could merge the PR as it is, and make those changes myself. If you would like to see what changes I have in mind, I can make specific comments. It's certainly nothing major, just a few tweaks. Let me know what you would prefer.

**Comment by torcado194** [comment]
Thanks for the review!

I would be interested in hearing what changes you have in mind :)

File: addons/better-terrain/BetterTerrain.gd [hunk 1]
```diff
@@ -583,6 +583,22 @@ func get_tile_terrain_type(td: TileData) -> int:
...
 
 
+## Returns an Array of all [TileData] tiles included in the specified
+## terrain [code]type[/code] for the [TileSet] [code]ts[/code]
+func get_tiles_in_terrain(ts: TileSet, type: int) -> Array[TileData]:
+	var result:Array[TileData] = []
+	
+	var cache := _get_cache(ts)
+	var tiles = cache[type]
+	if tiles and tiles.size() > 0:
```
Comment (Portponky) [inline addons/better-terrain/BetterTerrain.gd:593]: I would prefer
```
if !tiles:
    return result
```
to early return, and you don't need to check the size because no iteration will occur on an empty array

File: addons/better-terrain/BetterTerrain.gd [hunk 2]
```diff
@@ -708,6 +739,72 @@ func set_cells(tm: TileMap, layer: int, coords: Array, type: int) -> bool:
...
+	if type >= cache.size():
+		return false
+	
+	if cache[type].is_empty():
+		return false
+	
+	var changed = false
+	var potential_tiles = get_tiles_in_terrain(ts, type)
+	var td = tm.get_cell_tile_data(layer, coord)
+	if td:
```
Comment (Portponky) [inline addons/better-terrain/BetterTerrain.gd:762]: Similarly, early return with `if !td:` and `if place_peering.is_empty():`

File: addons/better-terrain/BetterTerrain.gd [hunk 3]
```diff
@@ -708,6 +739,72 @@ func set_cells(tm: TileMap, layer: int, coords: Array, type: int) -> bool:
...
+	if cache[type].is_empty():
+		return false
+	
+	var changed = false
+	var potential_tiles = get_tiles_in_terrain(ts, type)
+	for c in coords:
+		var td = tm.get_cell_tile_data(layer, c)
+		if td:
+			var placed_peering = tile_peering_for_type(td, type)
+			if not placed_peering.is_empty():
```
Comment (Portponky) [inline addons/better-terrain/BetterTerrain.gd:797]: Same here, these checks can be reversed and changed to `continue` statements.

File: addons/better-terrain/editor/Dock.gd [hunk 4]
```diff
@@ -350,30 +365,56 @@ func perform_edit_terrain(index: int, name: String, color: Color, type: int, cat
...
 
 
 func _on_fill_pressed() -> void:
 	draw_button.button_pressed = false
+	line_button.button_pressed = false
 	rectangle_button.button_pressed = false
 	fill_button.button_pressed = true
 
 
+func _on_replace_pressed() -> void:
```
Comment (Portponky) [inline addons/better-terrain/editor/Dock.gd:394]: This function appears to be unused, it can be removed (and if connected to a signal, unconnected).

File: addons/better-terrain/editor/Dock.gd [hunk 5]
```diff
@@ -480,20 +553,30 @@ func canvas_input(event: InputEvent) -> bool:
 			initial_click = current_position
 		var type = selected.get_index()
 		
-		if draw_button.button_pressed:
+		if paint_action == PaintAction.LINE:
+			pass
```
Comment (Portponky) [inline addons/better-terrain/editor/Dock.gd:557]: The way this is structured is correct, but it looks a little odd at a glance, so it could do with a comment explaining it. I figure you're separating out line drawing from the other conditions below?

File: addons/better-terrain/editor/Dock.gd [hunk 6]
```diff
@@ -510,3 +593,47 @@ func canvas_input(event: InputEvent) -> bool:
...
+	if event is InputEventKey:
+		if event.keycode == KEY_C and event.ctrl_pressed and not event.echo:
+			get_viewport().set_input_as_handled()
+			tile_view.copy_selection()
+		if event.keycode == KEY_V and event.ctrl_pressed and not event.echo:
+			get_viewport().set_input_as_handled()
+			tile_view.paste_selection()
+
+
+##bersenham alg ported from Geometry2D::bresenham_line()
```
Comment (Portponky) [inline addons/better-terrain/editor/Dock.gd:609]: This is a bit of an odd one, but Godot 4 gets really weird about formatting of ## comments. If you have even one that it doesn't like, it can prevent Godot from parsing scripts for custom comments, and it doesn't tell you why. I believe it requires a space after the double hash. Also, minor typo in algorithm name (should be bresenham).

File: addons/better-terrain/editor/TileView.gd [hunk 7]
```diff
@@ -2,6 +2,7 @@
 extends Control
 
 @onready var checkerboard := get_theme_icon("Checkerboard", "EditorIcons")
+@onready var dock = $"../../../../.."
```
Comment (Portponky) [inline addons/better-terrain/editor/TileView.gd:5]: Indexing into the parent node path is a bit inflexible. I would rather replace this with appropriate signals, which are connected by the dock. Looks like you would need two signals, for `paste_occured` and `change_zoom_level` or something along those lines.

File: addons/better-terrain/editor/TileView.gd [hunk 8]
```diff
@@ -166,7 +186,109 @@ func tile_part_from_position(position: Vector2i) -> Dictionary:
...
+				var target_rect := Rect2(offset + zoom_level * rect.position, zoom_level * rect.size)
+				if target_rect.has_point(position):
+					return target_rect
+			
+			offset.y += zoom_level * source.texture.get_height()
+	
+	return Rect2(-1,-1,0,0)
+
+
+#func get_tiles_in_rect(rect:Rect2i) -> Array[Vector2i]:
```
Comment (Portponky) [inline addons/better-terrain/editor/TileView.gd:234]: Remove debug comment.

File: addons/better-terrain/editor/TileView.gd [hunk 9]
```diff
@@ -166,7 +186,109 @@ func tile_part_from_position(position: Vector2i) -> Dictionary:
...
+				}
+				var pos = target_rect.position + target_rect.size/2
+				_build_tile_part_from_position(result, pos, target_rect)
+				tiles.push_back(result)
+			var alt_count := source.get_alternative_tiles_count(coord)
+			for a in alt_count:
+				var alt_id := 0
+				if a == 0:
+					continue
+				else:
```
Comment (Portponky) [inline addons/better-terrain/editor/TileView.gd:268]: Don't need to `else` after a `continue`

File: addons/better-terrain/editor/TileView.gd [hunk 10]
```diff
@@ -190,11 +312,12 @@ func _draw_tile_data(texture: Texture2D, rect: Rect2, src_rect: Rect2, td: TileD
 	if paint < 0 or paint >= BetterTerrain.terrain_count(tileset):
 		return
 	
-	var paint_terrain := BetterTerrain.get_terrain(tileset, paint)
-	for p in BetterTerrain.data.get_terrain_peering_cells(tileset, terrain.type):
-		if paint in BetterTerrain.tile_peering_types(td, p):
-			var side_polygon = BetterTerrain.data.peering_polygon(tileset, terrain.type, p)
-			draw_colored_polygon(transform * side_polygon, Color(paint_terrain.color, 0.6))
+	if draw_sides:
```
Comment (Portponky) [inline addons/better-terrain/editor/TileView.gd:315]: Early return rather than indent.

File: addons/better-terrain/editor/TileView.gd [hunk 11]
```diff
@@ -263,77 +396,250 @@ func _draw() -> void:
...
-					undo_manager.add_undo_method(BetterTerrain, &"set_tile_terrain_type", tileset, highlighted_tile_part.data, type)
-				undo_manager.add_undo_method(self, &"queue_redraw")
-				undo_manager.commit_action()
-		elif paint_action == PaintAction.DRAW_PEERING:
-			if highlighted_tile_part.has("peering"):
-				if !(paint in BetterTerrain.tile_peering_types(highlighted_tile_part.data, highlighted_tile_part.peering)):
-					undo_manager.create_action("Set tile terrain peering type", UndoRedo.MERGE_DISABLE, tileset)
-					undo_manager.add_do_method(BetterTerrain, &"add_tile_peering_type", tileset, highlighted_tile_part.data, highlighted_tile_part.peering, paint)
+		if clicked:
+			if event.button_index == MOUSE_BUTTON_LEFT:
```
Comment (Portponky) [inline addons/better-terrain/editor/TileView.gd:516]: There's a lot of indentation here, though it seems some may be unavoidable. I'd combine these if statements into
`if event.button_index == MOUSE_BUTTON_LEFT and taged_paste_tile_states.size() > 0:`

File: addons/better-terrain/editor/TileView.gd [hunk 12]
```diff
@@ -263,77 +396,250 @@ func _draw() -> void:
...
+						staged_rect.position -= base_rect.position + base_rect.size / 2
+						
+						staged_rect.position *= zoom_level
+						staged_rect.size *= zoom_level
+						
+						staged_rect.position += Vector2(current_position)
+						
+						var old_tile_part = tile_part_from_position(staged_rect.get_center())
+						var new_tile_state = p
+						if old_tile_part.valid and new_tile_state.part.valid:
```
Comment (Portponky) [inline addons/better-terrain/editor/TileView.gd:531]: I'd also flip this into a continue statement to remove another level of indentation.

File: addons/better-terrain/editor/TileView.gd [hunk 13]
```diff
@@ -263,77 +396,250 @@ func _draw() -> void:
...
+							for side in range(16): #not sure how to loop over or get max value of an enum like TileSet.CellNeighbor
```
Comment (Portponky) [inline addons/better-terrain/editor/TileView.gd:532]: The built in enums don't have a max, so range(16) is what I did somewhere else. Perhaps a `const MAX_CELL_NEIGHBOR_TYPES = 16` would be clearer, though I'm not sure it's really worthwhile.

File: addons/better-terrain/editor/TileView.gd [hunk 14]
```diff
@@ -263,77 +396,250 @@ func _draw() -> void:
...
+								var old_peering = BetterTerrain.tile_peering_types(old_tile_part.data, side)
+								var new_sides = new_tile_state.sides
+								if old_peering.has(paint) != new_sides.has(side):
+									if new_sides.has(side):
```
Comment (Portponky) [inline addons/better-terrain/editor/TileView.gd:536]: I'd use `if new_sides.has(side) and not old_peering.has(paint):` and for the else, `elif old_peering.has(paint) and not new_sides.has(side):`. That would be easier to read for me, and less indented.

File: addons/better-terrain/editor/Dock.gd (hunk 5 above)
Comment (torcado194) [reply addons/better-terrain/editor/Dock.gd:557]: Yeah, for line drawing mode execution happens on release, so this prevents other actions from triggering on press. i could alternatively wrap the other checks in an `if not paint_action == PaintAction.LINE` or return early. for now i've included a comment.

Comment (Portponky) [reply addons/better-terrain/editor/Dock.gd:557]: Yeah, nothing wrong with it, just thought it'd be good to improve the clarity.

File: addons/better-terrain/editor/Dock.gd (hunk 6 above)
Comment (torcado194) [reply addons/better-terrain/editor/Dock.gd]: whoops good catch

File: addons/better-terrain/editor/TileView.gd (hunk 7 above)
Comment (torcado194) [reply addons/better-terrain/editor/TileView.gd]: good idea, i havent used signals a whole ton. let me know if the new changes look good (it works fine)

File: addons/better-terrain/editor/TileView.gd (hunk 13 above)
Comment (torcado194) [reply addons/better-terrain/editor/TileView.gd]: hm that's annoying. ok that's fine, i've removed the comment for now

**Comment by torcado194** [comment]
These should all be fixed now :)
(apologies if resolving the github reviews myself wasn't the correct way to do it)

**Comment by Portponky** [comment]
That's great. I think you resolved the first few without supplying a patch though. If you don't get around to fixing those, I can fix them afterwards as it's just minor refactoring. I'll give the code a test with a bunch of different tilesets tomorrow and if it has no major issues, I will merge it. :+1:

**Comment by torcado194** [comment]
Oops my bad, I didn't commit the file. Should be good now.

Sounds good, thanks again! :)

**Comment by torcado194** [comment]
one more small update: painting/drawing lines now works correctly with other tile shapes

sx-899

File: addons/better-terrain/BetterTerrain.gd [hunk 15]
```diff
@@ -708,6 +740,80 @@ func set_cells(tm: TileMap, layer: int, coords: Array, type: int) -> bool:
 	return true
 
 
+## Replaces an existing tile on the [TileMap] for the [code]layer[/code]
+## and [code]coord[/code] with a new tile in the provided terrain [code]type[/code] 
+## in [TileSet] [code]type[/code] *only if* there is a tile with a matching set of
```
Comment (Portponky) [inline addons/better-terrain/BetterTerrain.gd:745]: `[TileSet] [code]ts[/code]`

**Comment by Portponky** [comment]
Been very busy the last couple days, sorry about that. I've been experimenting with the overwrite mode, and it doesn't seem to do what I expect it to do (or it has some issue?) I'll continue to investigate.

**Comment by torcado194** [comment]
No worries!
with overwrite mode, it only works on tiles within the terrain you're painting with that have matching peering sides. so like here, both terrains include both tilesets 



if that's not the case then there's probably an issue.

This behavior could very well be changed, it might be better to have "matching tiles" check for peering sides of any id, not matching ideas (so, any slope tile can replace any corner tiles, etc.)

The method I chose was mostly arbitrary, both would solve the usecase I mentioned in the original post.

**Comment by Portponky** [comment]
There's a few bugs in the overwrite code. The most important is that dictionary keys are not hashed in a stable manner; each dictionary gets a randomization to its hash function. That means that two dictionaries with the same keys will (potentially) have them in different orders.

The output of `tile_peering_for_type` was therefore in dictionary key order. This means that two tiles with the same peering bits might differ in order: [0, 8] compared to [8, 0]. As the replace cell(s) function does a direct comparison (which is good) this caused false negatives. This is easy to correct with a sort at the end:

```
	result.sort()
	return result
```

Another bug is that the replace cell(s) function skips tiles with no peering bits. However, it's totally possible that you would want to overwrite single tiles, so this skip should not occur.

Along with these additions, some speedups can be made, namely caching the peering types for the multi-cell function, and breaking when finding a match to prevent unnecessary checks. Also, eliminating some variables (GDScript, as a rule of thumb, tends to run faster with fewer lines of code, even if those lines are slightly denser). Here are my suggested functions:

```
## Replaces an existing tile on the [TileMap] for the [code]layer[/code]
## and [code]coord[/code] with a new tile in the provided terrain [code]type[/code] 
## in [TileSet] [code]ts[/code] *only if* there is a tile with a matching set of 
## peering sides in this terrain.
## Returns [code]true[/code] if any tiles were changed. Use [method replace_cells]
## to replace multiple tiles at once.
func replace_cell(tm: TileMap, layer: int, coord: Vector2i, ts: TileSet, type: int) -> bool:
	if !tm or !tm.tile_set or layer < 0 or layer >= tm.get_layers_count() or type < 0:
		return false
	
	var cache := _get_cache(tm.tile_set)
	if type >= cache.size():
		return false
	
	if cache[type].is_empty():
		return false
	
	var td = tm.get_cell_tile_data(layer, coord)
	if !td:
		return false
	
	var placed_peering = tile_peering_for_type(td, type)
	for pt in get_tiles_in_terrain(ts, type):
		var check_peering = tile_peering_for_type(pt, type)
		if placed_peering == check_peering:
			var tile = cache[type].front()
			tm.set_cell(layer, coord, tile[0], tile[1], tile[2])
			return true
	
	return false


## Replaces existing tiles on the [TileMap] for the [code]layer[/code]
## and [code]coords[/code] with new tiles in the provided terrain [code]type[/code] 
## in [TileSet] [code]ts[/code] *only if* there is a tile with a matching set of 
## peering sides in this terrain for each tile.
## Returns [code]true[/code] if any tiles were changed.
func replace_cells(tm: TileMap, layer: int, coords: Array, ts: TileSet, type: int) -> bool:
	if !tm or !tm.tile_set or layer < 0 or layer >= tm.get_layers_count() or type < 0:
		return false
	
	var cache := _get_cache(tm.tile_set)
	if type >= cache.size():
		return false
	
	if cache[type].is_empty():
		return false
	
	var changed = false
	var potential_peering = get_tiles_in_terrain(ts, type).map(
		func(tile): return tile_peering_for_type(tile, type)
	)
	for c in coords:
		var td = tm.get_cell_tile_data(layer, c)
		if !td:
			continue
		
		var placed_peering = tile_peering_for_type(td, type)
		for check_peering in potential_peering:
			if placed_peering == check_peering:
				var tile = cache[type].front()
				tm.set_cell(layer, c, tile[0], tile[1], tile[2])
				changed = true
				break
	
	return changed
```

With these changes, the overwrite appears to be working exactly as you describe. If you are happy with this, feel free to update the PR. I'm going to test out the other edge cases in the meantime. Thanks!

**Comment by univeous** [comment]
These new features look great. Personally, I'd want to go beyond just copying and pasting peering data and have the ability to save them somewhere so we can summon them from the void. 
This would make it very easy to set up new peering data on different tilesets.

**Comment by Portponky** [comment]
Overwrite mode doesn't work too well with types defined by categories :thinking: so I will have a think about an improvement in that regard.

**Comment by torcado194** [comment]
Oh, thanks for spotting those! Yeah without knowing how the internal structures were handled I was just making my best guess referencing the other functions.

You're right about not skipping tiles with no peering sides. It feels a little weird to me, I guess because I'm mentally considering the "painted tiles" in the terrain as potential candidates for what i want to replace, and this change opens up the whole tileset for that check. But I think it's a correct change, I can't come up with any reasons why it shouldn't be that way.

Yeah, categories make it tricky. While typing up my thought process I ended up implementing a fix anyway. The new push has changes in `replace_cell(s)` so that it now includes all categories the provided terrain is in for its match checks (and conceptually it would check against categories the placed tile is in, but if they aren't part of the same category then they shouldn't match anyway, so it only collects categories from the provided terrain)

I wasn't sure how to best integrate your `replace_cells` changes into the new code, or if it even needs it now. Let me know your thoughts on that, and on the new process in general.

**Comment by torcado194** [comment]
> These new features look great. Personally, I'd want to go beyond just copying and pasting peering data and have the ability to save them somewhere so we can summon them from the void. This would make it very easy to set up new peering data on different tilesets.

When first thinking about how to implement this, I was considering how to serialize this data to store in the clipboard. But I tested other parts of the editor and discovered that they also just use internal memory and don't store anything to the clipboard (e.g. copying nodes in the tree), so I just followed that convention. That serialization would be necessary for saving out terrains, I'd have to think about it more. Maybe Portponky would have a better idea of how to handle that.

**Comment by univeous** [comment]
> When first thinking about how to implement this, I was considering how to serialize this data to store in the clipboard. But I tested other parts of the editor and discovered that they also just use internal memory and don't store anything to the clipboard (e.g. copying nodes in the tree), so I just followed that convention. That serialization would be necessary for saving out terrains, I'd have to think about it more. Maybe Portponky would have a better idea of how to handle that.

Sounds reasonable. After some thought, I don't think the feature I mentioned is as important as I thought it was before. Of course being able to have that feature would still make people's lives a little easier. My other complaint is that we should make the default function of the scroll wheel zoom, and only scroll up and down when holding ctrl, which would be consistent with the default tileset and tilemap panel.

**Comment by Portponky** [comment]
> My other complaint is that we should make the default function of the scroll wheel zoom, and only scroll up and down when holding ctrl, which would be consistent with the default tileset and tilemap panel.

Unfortunately this makes it pretty unusable on a laptop touchpad. Most other apps use ctrl as a zoom modifier; it is the default panels that are wrong here imo.

**Comment by Portponky** [comment]
This looks pretty good overall, so I will try to merge it this evening! Great work.

**Comment by torcado194** [comment]
Great, thanks a lot! :)

**Review by Portponky — APPROVED** [review]
